sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/RequiresStableInputIT.java

```diff
@@ -0,0 +1,167 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.nio.charset.Charset;
+import java.util.Date;
+import java.util.UUID;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.options.Validation.Required;
+import org.apache.beam.sdk.testing.FileChecksumMatcher;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.testing.TestPipelineOptions;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Integration test for the support of {@link
+ * org.apache.beam.sdk.transforms.DoFn.RequiresStableInput} annotation.
+ */
+@RunWith(JUnit4.class)
+public class RequiresStableInputIT {
+
+  private static final String VALUE = "value";
+  // SHA-1 hash of string "value"
+  private static final String VALUE_CHECKSUM = "f32b67c7e26342af42efabc674d441dca0a281c5";
+
+  private static class PairWithRandomKeyFn extends SimpleFunction<String, KV<String, String>> {
+    @Override
+    public KV<String, String> apply(String value) {
+      String key = UUID.randomUUID().toString();
+      return KV.of(key, value);
+    }
+  }
+
+  private static class MakeSideEffectAndThenFailFn extends DoFn<KV<String, String>, String> {
+    private final String outputPrefix;
+
+    private MakeSideEffectAndThenFailFn(String outputPrefix) {
+      this.outputPrefix = outputPrefix;
+    }
+
+    @RequiresStableInput
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      MatchResult matchResult = FileSystems.match(outputPrefix + "*");
+      boolean firstTime = (matchResult.metadata().size() == 0);
+
+      KV<String, String> kv = c.element();
+      writeTextToFileSideEffect(kv.getValue(), outputPrefix + kv.getKey());
+      if (firstTime) {
+        throw new Exception("Deliberate failure: should happen only once.");
```

Comment:
It will happen twice, once for Multi and once for Single.

`should happen only once` -> `should happen only once for each application of the DoFn within the transform graph.`